Webセキュリティ

1mori opened the pull request:
## 関連ページ
https://bootcamp.fjord.jp/practices/231

## 最終更新日
2024/06/26

## やりたいこと
- 脆弱性のあるRailsアプリケーションを修正する

## やったこと
- 位置指定ハンドラを用い、文字列をサニタイズした
- HTMLエスケープ後、改行文字に変換をすることでブログ内でスクリプトを実行できなくした
- ブログの編集、更新、削除にユーザー制限を設定した
- 非公開ブログに対してユーザー制限を設定した、非ログインユーザーに対しても同様に設定
- 一般ユーザーに対して`rondom_eyecatch`の権限を付与しないようにした
- APIで返ってくるユーザー情報を`nickname`に制限した

1mori committed:
fix: 検索フォームのSQL脆弱性に対応

`sanitize_sql_like`という、LIKE検索のワイルドカードをエスケープしてくれるメソッドを使用
検索文内に変数は直接書かず、?を置いて後から代入する

1mori committed:
add: 検索しないときにが動かないように条件文を追加

1mori committed:
fix: XSSを起こさないようにformat_contentの処理を修正

`h()`でタグを残し、`simple_format()`で改行を残すようにした

1mori committed:
fix: URL直打ちに対する脆弱性の改修

他人の作成したブログの編集画面に飛べないように制限を設けた
編集・更新・削除に対して制限を作成した

他人の非公開ブログに対してアクセス制限を設けた
非公開ブログかつ、未ログイン・または他ユーザーの場合はエラーを出す

1mori committed:
fix: APIで送信されるデータを制限

少なくとも今回はメールアドレスを送信してはならないため、`email`送信項目から削除した

1mori committed:
fix: プレミアムユーザにのみrondom_eyecatchパラメータを付与する

1mori committed:
delete: jsonファイル内の他に必要のない要素も削除

今回送るべき情報は「いいねした人」として表示される`nickname`のみである
課題のテストコードとしては`email`を削除すればOKだが、他の情報も必要はないと思い削除した

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -48,6 +50,16 @@ def set_blog
   end
 
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    allow_params = %i[title content secret]
+    allow_params << :random_eyecatch if current_user.premium?
+    params.require(:blog).permit(allow_params)
```
変更前後で `permit` メソッドに渡す引数のデータ型が変わってしまっています。

1mori replied:
スプラット演算子を用いて中身を展開してから渡すようにしました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -48,6 +50,16 @@ def set_blog
+    params.require(:blog).permit(allow_params)
+  end
+
+  def match_login_user
+    raise ActiveRecord::RecordNotFound if @blog.user_id != current_user.id
+  end
+
+  def show_secret
+    raise ActiveRecord::RecordNotFound if @blog.secret && (current_user.nil? || @blog.user_id != current_user.id)
   end
```
`ActiveRecord::RecordNotFound` は基本的に Rails が扱う例外なので、自分で raise するのではなく Rails が送出するような実装にしてください 🙏

1mori replied:
`find_by!`を用いて見つからないときは例外を出すように書き換えました。

cafedomancer commented on app/models/blog.rb:
```diff
@@ -10,7 +10,12 @@ class Blog < ApplicationRecord
 
   scope :search, lambda { |term|
-    where("title LIKE '%#{term}%' OR content LIKE '%#{term}%'")
+    if term.present?
+      sanitized_term = sanitize_sql_like(term)
+      where('title LIKE ? OR content LIKE ?', "%#{sanitized_term}%", "%#{sanitized_term}%")
+    else
+      all
+    end
```
引数の `term` に `nil` を渡した時に発行されるクエリが変更前後で変わってしまっています。

1mori replied:
後ろに`.to_s`をつけて同じクエリ文を発行できるようにしました。

cafedomancer commented on app/views/api/liking_users/_user.json.jbuilder:
```diff
@@ -1 +1 @@
-json.extract! user, :id, :email, :created_at, :updated_at, :nickname, :premium
+json.extract! user, :nickname
```
アプリケーション中で使われている必要な情報も消えてしまっています。

1mori replied:
idは必要な情報でした。変更しました。

1mori committed:
fix: 渡す引数にデータ型を買えないように修正

スプラット演算子(*)を用いてリストの中身を展開して渡すようにした

1mori committed:
change: ユーザー一致確認の際、Railsにエラーを出させるようにした

1mori committed:
change: idをAPIに載せるように変更

1mori committed:
add: to_sをつけてnilに対応した

1mori committed:
change: 隠しブログに対する権限の文章を変更

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -48,6 +50,20 @@ def set_blog
 
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    allow_params = %i[title content secret]
```
変数名は名詞にしてください :pray: allow は動詞なので、名詞と組み合わせても名詞にはならないです。

1mori replied:
`allowed_params`に変更しました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -48,6 +50,20 @@ def set_blog
+  end
+
+  def match_login_user
+    @blog = Blog.find_by!(id: params[:id], user_id: current_user.id)
```
`set_blog` と合わせて Blog 検索が 2 回行われます。データベースへの問い合わせは時間のかかる処理なので、Blog の検索は 1 回で済ませてください。`show_secret` についても同様です。

1mori replied:
指摘ありがとうございます。
`match_login_user`ではedit, update, destroyのDB検索を、`show_secret`ではshowのDB検索をしており、`set_blog`の処理が必要ないことがわかったので削除しました。

cafedomancer commented on app/models/blog.rb:
```diff
@@ -10,7 +10,12 @@ class Blog < ApplicationRecord
 
   scope :search, lambda { |term|
-    where("title LIKE '%#{term}%' OR content LIKE '%#{term}%'")
+    if term.present?
+      sanitized_term = sanitize_sql_like(term.to_s)
+      where('title LIKE ? OR content LIKE ?', "%#{sanitized_term}%", "%#{sanitized_term}%")
+    else
+      all
+    end
```
https://github.com/1mori/haunted-blog/pull/1#discussion_r1656504262 では `term` が `nil` のとき、つまり else 節が実行されるときについてコメントしているので、問題が解決していなさそうに見えました。

1mori replied:
前回`.to_s`としたことで`nil`が空文字列に変換されるようにしたにも関わらず、必要のなくなったif~elseを消していませんでした。
条件分岐は削除し、`nil`が入力された場合でも問題なく実行できることを確認しました。

1mori committed:
delete: set_blogを削除

match_login_userではedit, update, deleteのDB検索を、
show_secretではshowのDB検索を行っており、set_blogは重複する処理であったため、削除した

1mori committed:
change: allow_params -> params

1mori committed:
fix: nilが入力された時でも元の処理と同様になるように処理を変更

前回入力された文章を`.to_s`で文字列に変換するようにしていたため、
elseの分岐は必要なくなっていた、なのでif文を削除した

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,11 +44,21 @@ def destroy
 
-  def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+  def match_login_user
+    @blog = Blog.find_by!(id: params[:id], user_id: current_user.id)
+  end
+
+  def show_secret
+    @blog = if current_user
+              Blog.where(id: params[:id]).where('secret = ? OR user_id = ?', false, current_user.id).first!
+            else
+              Blog.find_by!(id: params[:id], secret: false)
+            end
   end
```
Blog の検索とインスタンス変数への設定をこのあたりのメソッドで行うのはよいと思うのですが、元々この作業を行なっていた `set_blog` というメソッドとは名前がかけ離れてしまっています。やっていることがメソッド名として表現できていない気がするので、名前を見直した方が良さそうな気がしました 👀

1mori replied:
`match_login_user`を`set_blog`に改名しました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,7 +3,8 @@
 
-  before_action :set_blog, only: %i[show edit update destroy]
+  before_action :match_login_user, only: %i[edit update destroy]
+  before_action :show_secret, only: %i[show]
```
フィルターはアクション間で処理を共通化したり、本質的でない処理をアクションから切り離すために使われるものなので、特定のアクションでしか実行されない本質的な処理をフィルターで実行する必要はないと思います。

1mori replied:
指摘ありがとうございます。
showswのDB検索は`def show`内に書くようにしました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,11 +44,21 @@ def destroy
-  def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+  def match_login_user
+    @blog = Blog.find_by!(id: params[:id], user_id: current_user.id)
```
特定のユーザーに紐づくデータを参照する場合は、必ずユーザーの関連からデータをたどるようにしてください 🙏 `Blog.xxx` からたどる方式だと、`user_id: ...` という条件の指定し忘れによって情報漏洩に簡単につながってしまうからです。

1mori replied:
`current_user.blog.`という書き方にし、辿る順を変更しました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,11 +44,21 @@ def destroy
+
+  def show_secret
+    @blog = if current_user
+              Blog.where(id: params[:id]).where('secret = ? OR user_id = ?', false, current_user.id).first!
```
検索の意味合い的には「`'secret = ? OR user_id = ?', false, current_user.id` という条件にマッチするブログのうち、ID が `params[:id]` に合致するもの」という感じかなーと思います。今の書き方だとそれが表現されていなさそうです。else の方の処理についても同様です。

1mori replied:
Discordでもコメントありがとうございます。
whereで条件を絞り、findで合致するIDのブログを取得する書き方に変更しました。

1mori committed:
change: match_login_user -> set_blog

1mori committed:
change: show_secretを削除し、中身をshowに移動

show_secretの実行結果は`show`でのみ使用しているので、メソッドは削除し、
def show ないで中身を実行することにした

1mori committed:
change: edit, update, destroy権限に対しての確認方法を変更

Blogのuser_idからではなく、current_user側から辿るように変更した

1mori committed:
fix: 秘密のブログが見えないようにするクエリ文を変更

whereで条件検索をし、最後にfindで一致するIDを検索するように動作を変更した

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,11 +49,13 @@ def destroy
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+  def set_blog
+    @blog = current_user.blogs.find(params[:id])
   end
```
メソッドの定義順を入れ替えると差分が見づらくなるので入れ替えないほうがいいかなーと思います。

1mori replied:
match_login_userから名称を変更した関係上、入れ替わっていることを失念していました。
指摘ありがとうございます。修正しました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,19 @@
+    @blog = if current_user
+              Blog.where('secret = ? OR user_id = ?', false, current_user.id).find(params[:id])
+            else
+              Blog.where(secret: false).find(params[:id])
```
`where(secret: false)` の部分は既存の実装が使えるはずです。

1mori replied:
modelを見ていたところ、:publishedに`secret: false`の記載があることに気づいたので、`Blog.published`に変更しました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,19 @@
-  def show; end
+  def show
+    @blog = if current_user
+              Blog.where('secret = ? OR user_id = ?', false, current_user.id).find(params[:id])
```
where の OR の部分については Active Record の API を使って表現してもいいかなと思います。そうすれば既存の実装と組み合わせて使えるからです。

1mori replied:
`.or`を用いた書き方に変更しました。
また、先ほどのpublishedを用いて`current_user`の時にクエリ文にorの条件を付与するようにし、else以降は削除しました。

1mori committed:
change: クエリ文の書き換え

共通条件であるsecret: falseのクエリを最初に実行し、その後、ユーザーが秘密のブログ作成者本人であれば表示するという書き方に変更
最後に`.find(params[:id])`とすることでブログのID検索ができるようにした

1mori committed:
change: 入れ替わっている箇所を変更

1mori committed:
change: publishedを用いた書き方に変更した

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,18 @@
 
-  def show; end
+  def show
+    query = Blog.published
+    query = query.or(Blog.where(user_id: current_user.id)) if current_user
+
+    @blog = query.find(params[:id])
```
これはちょっと読みづらいですね... 😵 16 行目の `query.find(params[:id])` の部分で、最終的にどのようなデータ集合から検索が行われるのか、コードを注意深く読まないと分からないからです。if else を使う元の形のほうが分かりやすいと思います。

1mori replied:
コメントありがとうございます。
確かに検索が一行で収まっている方がわかりやすいですね。形式は元の`if~else`に戻しておきました。

1mori committed:
change: クエリ文を元のif~elseの形に戻した

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,19 @@
-  def show; end
+  def show
+    @blog = if current_user
+              Blog.published.or(Blog.where(user_id: current_user.id)).find(params[:id])
```
where に関連を指定する場合は、ID ではなくオブジェクトをそのまま指定するほうが読みやすいかなーと思います。
